### PR TITLE
Airport Challenge-James-Mackintosh

### PR DESCRIPTION
**Pull request by jazmac**

### CODE REVIEW

**Review by jazmac — COMMENTED** [review]
I could've done a lot more edge case testing in my airport.spec.js file to ensure that unexpected values wouldn't be accepted.

File: tests/airport.spec.js [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+let expected;
+let actual;
+let result;
+let airport;
+let plane;
+let plane2;
+let plane3;
+let plane4;
+let plane5;
+let plane6;
```
Comment (jazmac) [inline tests/airport.spec.js:14]: I still haven't fully grasped how the instance of the plane class can work here so there was likely a better way of testing multiple instances of the plane class without creating separate planes like this for test 6.

File: tests/airport.spec.js [hunk 2]
```diff
@@ -149,13 +154,22 @@ console.log(`Test 6 - Check that you can't add planes beyond set airport capacit
...
+plane4 = new Plane();
+plane5 = new Plane();
+plane6 = new Plane();
 expected = 5;
 
 // Act
 airport.setCapacity(5);
-for (let i = 0; i < 6; i++) {
-    airport.landPlane(plane);
-}
```
Comment (jazmac) [inline tests/airport.spec.js]: I honestly wasn't happy switching away from the for loop when refactoring this test that I thought was an elegant way of running my landplane method, but I had a difficult time wrapping my head around being versatile in how I call instances of classes.

File: src/airport.js [hunk 3]
```diff
@@ -1,6 +1,44 @@
...
 
-}
+    constructor(maximumCapacity = 10) {
+        this.airportPlanes = [];
+        this.maximumCapacity = maximumCapacity;
+    };
+
+    landPlane = plane => {
+        if (this.airportPlanes.length >= this.maximumCapacity) {
+            return this.airportPlanes.length;
```
Comment (jazmac) [inline src/airport.js:12]: I know this should've been something like a boolean return instead of just returning the array size. I'm not sure why I chose to frame my tests around the size of the array from this return like this.

File: src/airport.js [hunk 4]
```diff
@@ -1,6 +1,44 @@
...
+    setCapacity = (input) => {
+        this.maximumCapacity = input;
+    };
+
+    planeAtAirport = plane => {
+        if (this.airportPlanes.includes(plane)) {
+            return true;
+        }
+        else {
+            return false;
```
Comment (jazmac) [inline src/airport.js:37]: This applies to many of the methods but I think just return a true/false boolean could be improved upon. Perhaps catching errors and returning a message to the user when invalid values or when a plane is already landed at the airport or when a plane that doesn't exist at the airport tries to take off.

File: src/airport.js [hunk 5]
```diff
@@ -1,6 +1,44 @@
...
+        if (this.airportPlanes.length >= this.maximumCapacity) {
+            return this.airportPlanes.length;
+        }
+        else if (plane instanceof Plane && !this.planeAtAirport(plane)) {
+            this.airportPlanes.push(plane);
+        };
+
+    };
+
+    takeoffPlane = plane => {
```
Comment (jazmac) [inline src/airport.js:20]: This method I had the most trouble understanding properly. I spent longer than I'd like to admit trying to understand exactly how values were being passed around and/or assigned. I don't think I would've naturally done it this way if similar methods in the basket challenge hadn't been shown to us in this form because it ended up impeding my understanding for a while.

File: src/airport.js [hunk 6]
```diff
@@ -1,6 +1,44 @@
...
+        const indexOfPlaneInAirportPlanes = this.airportPlanes.findIndex(planeManifest => planeManifest.id === plane.id);
+
+        if (indexOfPlaneInAirportPlanes > -1 && this.planeAtAirport(plane)) {
+            this.airportPlanes.splice(indexOfPlaneInAirportPlanes, 1);
+        };
+    };
+
```
Comment (jazmac) [inline src/airport.js:27]: This method I had the most trouble understanding properly. I spent longer than I'd like to admit trying to understand exactly how values were being passed around and/or assigned. I don't think I would've naturally done it this way if similar methods in the basket challenge hadn't been shown to us in this form because it ended up impeding my understanding for a while.

File: airport-domain-model.md [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+2. Override the maximum airport capacity to a different size limit using a method
+3. Check that more planes cannot be added beyond set airport capacity
+
+| Objects | Properties                            | Messages             | Outputs |
+| ------- | -------------------------             | -----------------    | ------- |
+| Airport | airportPlanes @Array[@Plane]          | landPlane(@Plane)    | @Void   |
+|         | MaximumCapacity @integer              | maximumCapacity()    | @Void   |
+|         |                                       |                      |         |
+| Plane   | id @String                            | getId()              | @String |
+
```
Comment (jazmac) [inline airport-domain-model.md:49]: I ended up combining these two user stories as while writing the code for setting the capacity, I inadvertently covered the 3rd User story criteria as well.

File: testing-framework.js [hunk 8]
```diff
@@ -0,0 +1,10 @@
+const assertEquals = (expectedOutput, actualOutput) => expectedOutput === actualOutput;
+const assertLessThan = (expectedOutput, actualOutput) => expectedOutput < actualOutput;
+const assertTrue = actualOutput => actualOutput === true;
+const assertFalse = actualOutput => actualOutput === false;
+const assertNotEquals = (expectedOutput, actualOutput) => expectedOutput !== actualOutput;
+
```
Comment (jazmac) [inline testing-framework.js:6]: I should've implemented other types of testing as I only used the assertEquals for all my testing purposes.

File: tests/airport.spec.js [hunk 9]
```diff
@@ -153,7 +153,7 @@ expected = 5;
 
 // Act
 airport.setCapacity(5);
-for (let i = 0; i <= airport.maximumCapacity; i++) {
```
Comment (jazmac) [inline tests/airport.spec.js]: This was foolish as all this would've done is run the test to the maximum capacity set and not actually try and push a plane onto the airport array beyond the set capacity. I realised this and amended the test.
